ntex-codec/src/encoder.rs

```diff
@@ -1,26 +1,22 @@
-use ntex_bytes::BytesMut;
-use std::rc::Rc;
+use ntex_bytes::{BufferSource, BytesPool};
+use std::future::Future;
 
 /// Trait of helper objects to write out messages as bytes.
-pub trait Encoder {
+pub trait Encoder<P = BytesPool> where P: BufferSource {
     /// The type of items consumed by the `Encoder`
     type Item;
 
     /// The type of encoding errors.
     type Error: std::fmt::Debug;
 
-    /// Encodes a frame into the buffer provided.
-    fn encode(&self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error>;
-}
-
-impl<T> Encoder for Rc<T>
-where
-    T: Encoder,
-{
-    type Item = T::Item;
-    type Error = T::Error;
+    /// The type of future returned by `Encoder::encode`.
+    type EncodeFuture: Future<Output = Result<(), Self::Error>>;
```

Comment:
I don't like idea of async on codec level. 
I see two options here, introduce new method, `encode_size(item: &Self::Item)` so top level code could
allocate enough storage, and then `encode` could accept `&mut P::Owned`. or encode could
return custom type so it can indicated required memory cap

Reply:
Yes, I considered having `encode` return `enum EncodeResult { Ok(()), InsufficientBuffer(usize), Err(Self::Error) }`, and having all callers retry it with larger `P::Owned` buffer if it fails with `InsufficientBuffer`, but the problem is that it's hard to roll back the changes to the `P::Accumulator` after discovering in the middle of an encode that the given `P::Owned` is insufficient size.

`encode_size` can work, though doing it at the `Encoder` level means every impl of `Encoder` has to support it. Also, note that `encode_size` will need to work in a perhaps unexpected way - it will only count bytes that are not managed by any `P::Shared` that its real `encode` will use with `put_bytes` directly. Eg: If you have a frame that's `struct Item { x: [u8; X], y: Shared<P>, z: [u8; z]`, where `y.len() == Y`, then `encode_size` would return `X + Z`, not `X + Y + Z`, and `encode` would do:

1. Verify that the `P::Owned` has `unfilled_len() > X + Z`.
1. Copy X bytes from `item.x` into `P::Owned`, split off the front X bytes as `x`, freeze into a `Shared`, and call `try_put_bytes(x)`
1. Call `try_put_bytes(item.y)`
1. Copy Z bytes from `item.z` into `P::Owned`, split off the front Z bytes as `z`, freeze into a `Shared`, and call `try_put_bytes(z)`

Reply:
ok, what about passing &mut Context instead. then it would be possible to use codec in future impls.

Reply:
What do you mean? Like:

```rust
    fn encode(&mut self, item: Self::Item, pool: &P, dst: &mut P::Accumulator, cx: &mut Context<'_>) -> Self::EncodeFuture;
```

? That cannot work, because BufferPool's `TakeFuture` will be dropped inside the function. BufferPool does not have magical connection to I/O reactor like tokio I/O streams do. It needs to hold on to context wakers itself, which means it has to ensure the task that polled the future still exists at the time it's woken up, which means it needs a `Future` to be held by the task.

Reply:
but that is not true, you dont need future to be waken up. future is just facade for waker. just store waker and call wake whenever memory is available

Reply:
it has nothing to do with tokio

Reply:
again, i am not insisting, you may have your reasons, but this api in this form not going to be part of ntex

Reply:
That is not correct.

Consider the case where the buffer is currently out of buffers. task1 calls `pool.poll_take(cx)` - this returns `Pending` and the pool stores `cx.waker().clone()` in its waker queue to call later when a buffer is returned.

task2 also calls `pool.poll_take(cx)` - same thing happens. The waker queue now has waker of task1 and waker of task 2.

Now task1 exits, without ever having waited for the buffer to be given to it.

Now a buffer is returned to the pool from somewhere. The pool pops the first waker, wask1's and `wake()`s it. But this is a no-op at best because the corresponding task is already gone. Pool has no way of knowing this, so it has no reason to pop another waker (task2's) and `wake()` that one up.

That's why you need a `Future` to be held by the task1, and wire it up so that the `Drop` of the `Future` also removes the corresponding waker from the pool's waker queue. When task1 exits, the `Future` it was holding on to is also dropped.

It's the same reason why, say, [`futures_util::Mutex::lock`](https://docs.rs/futures-util/0.3.17/futures_util/lock/struct.Mutex.html#method.lock) returns a Future, and you can see that the Future's Drop is wired up to remove the waker.

Reply:
>again, i am not insisting, you may have your reasons, but this api in this form not going to be part of ntex

Sure. And then you suggested `encode_size`, and I pointed out that will be fine with a caveat. You didn't respond to whether that caveat was acceptable for you or not.

Reply:
>it has nothing to do with tokio

I mentioned tokio as an example where you do not have to take this consideration in account, because something like a `tokio::io::TcpStream` does not have an internal queue of wakers that it can spontaneously decide to wake up later. Instead the tokio runtime and I/O reactor together work to mark the task that polled the `TcpStream` as something that should be woken up when an event on the `TcpStream` happens in the future. Thus it's trivially true that when such a task exits it should not be woken up for I/O events on that `TcpStream`, and no wakers are involved.

When you have something that *has* to manage the wakers itself, like the buffer pool, is when this consideration comes into play.